### ISSUE TITLE
Button event handler stuck on a non existing row

### ISSUE THREAD

**Issue by rtrzebinski** [maxence-charriere/go-app#826]
I have a table with 4 rows. Each has a `delete` button that deletes a row from the UI.

When deleting in DESC order: 4 -> 3 -> 2 -> 1 - all works fine.

When deleting in ASC order: 1 -> 2 -> 3 -> 4 - only first row is deleted.

By looking at the console logs I can see that the [event handler](https://github.com/rtrzebinski/simple-memorizer-4/blob/49b769d32146a988ce558331f243ea11fe10abeb/internal/frontend/components/exercise_row.go#L38) on `app.Button().Text("Delete").OnClick` is still trying to delete row with Id: 1, even though this row does not exist anymore.

**I wonder if this is some issue with event handlers not being refreshed correctly or am I doing something wrong.**

Note that API calls are not implemented, this is only about the UI.

Also please note that slices are indexed from 1 - I built the indexes to follow IDs from the database (starting from 1), so it is not a 0 index related problem.

Exercise component (parent) -> https://github.com/rtrzebinski/simple-memorizer-4/blob/49b769d32146a988ce558331f243ea11fe10abeb/internal/frontend/components/exercises.go

Exercise row component (children, nested) -> https://github.com/rtrzebinski/simple-memorizer-4/blob/49b769d32146a988ce558331f243ea11fe10abeb/internal/frontend/components/exercise_row.go

```go
			app.Button().Text("Delete").OnClick(func(ctx app.Context, e app.Event) {
				app.Logf("delete %d", h.exercise.Id)
				app.Logf("delete %d", h.exercise.Question)
				err := h.api.DeleteExercise(h.exercise)
				if err != nil {
					app.Log(fmt.Errorf("failed to delete exercise: %w", err))
				}
				// remove deleted row from the parent slice
				rows := make([]*ExerciseRow, len(h.parent.rows))
				for i, row := range h.parent.rows {
					if i != h.exercise.Id && row != nil {
						app.Log("inserting", i, row.exercise.Id, row.exercise.Question)
						rows[i] = row
					}
				}
				app.Log(h.parent.rows)
				h.parent.rows = rows
				app.Log(rows)
			}),
```

### Steps to reproduce

```
git clone git@github.com:rtrzebinski/simple-memorizer-4.git
git checkout 49b769d32146a988ce558331f243ea11fe10abeb
make dev
make run
open http://localhost:8000/exercises
```

Delete row 1 - will work.

Delete row 2 (that became row 1, since previous was deleted) - will not work.

### Actual result

See console logs - it says that `OnClick` event handler on the button in row 2 is still trying to delete non existing row 1 (Germany):

```
delete 1
delete %!d(string=Germany)
```

### Expected result

Rows are being deleted in any order

**Comment by oderwat** [comment]
I am uncertain if it is related, but the problem sounds familiar. Check what we were writing about it here #780 and here #759.

**Comment by rtrzebinski** [comment]
@oderwat You are my hero :D

I checked threads you shared and the hint was in setting the **button id** and then getting it with:

```go
ctx.JSSrc().Get("id").String()
```

This will be actually correct and rows are now being deleted in any order, thank you so much.

Commit fot the reference -> https://github.com/rtrzebinski/simple-memorizer-4/commit/b6b67dbf037b2f36fd25b2d4bdb4474e639748dd

Please let me know whether I should close this thread or not.

**Comment by oderwat** [comment]
I dislike using IDs and I skip as much JavaScript / DOM interactions as possible. Did you try using `OnClick(func().., fmt.Sprintf("%p", ExerciseRow))` instead?

**Comment by rtrzebinski** [comment]
@oderwat 

> I dislike using IDs and I skip as much JavaScript / DOM interactions as possible.

Me too, one of the main reasons for me to go with `go-app` is to **not** use the JS to build the UI.

> Did you try using OnClick(func().., fmt.Sprintf("%p", ExerciseRow)) instead?

Yes, this is what is not working correctly and why I needed to work around with the "JS get id" trick.

See the code in my first post, you can try to follow steps to reproduce on your machine.

**Comment by oderwat** [comment]
I "never" write a function inside the `OnClick()` call, but always use a function generator instead, that can use a closure to capture the ID.

**Comment by rtrzebinski** [comment]
@oderwat Thank you for an interesting suggestion.

I tried your approach to create a closure to capture the ID - with and without `id:=idArg` - sadly none of these helped, the same issue remains.

So the very only working solution appears to be using `ctx.JSSrc().Get("id").String()` so far.

**Comment by oderwat** [comment]
I had some time and got your code to run, after finding out that one needs the "golang-migrate" command installed.

So, I could try it myself and added a [PR](https://github.com/rtrzebinski/simple-memorizer-4/pull/3) that should fix it "legally". I also tried to explain why it does and what goes wrong without.

I think this is pretty obscure and wished that Go-App would not (constantly) let us trip over such problems.

**Comment by oderwat** [comment]
To leave that here:

The solution was to add the address of the component as scope to the `OnClick()` handler. This needs to be done like this: `OnClick(handlerFunc, fmt.Sprintf("%p",component))`. So that the code that checks for updates of the handler uses this pointer as scope for checking the equality  https://github.com/maxence-charriere/go-app/blob/db6035c6fa1badc265f12dce1e8494cfeee8eef0/pkg/app/event.go#L34-L50 and replaces the former handler with the one from the new "Render()".
Without that, the Go-App update will update all the node elements that changed but not the attached handler, which will stay the same as before.

Using the "%p" is needed because the scope will get converted to a string using `fmt.Sprint()` and this will dump the "content" of the struct otherwise. 

See also my (declined) PR https://github.com/maxence-charriere/go-app/pull/749 which was meant to change this behavior. I still think that this should get changed somehow. If not at `toString()` then earlier in the `toPath()` implementation (tagging you here @maxence-charriere, maybe this real-world case can serve as a heads-up on that "problem"?).

**Comment by rtrzebinski** [comment]
Fixed by https://github.com/rtrzebinski/simple-memorizer-4/pull/3/commits/9208038fb8bda58d47e5f2b9a9a9fb71cd9bf896

**Comment by rtrzebinski** [comment]
Re opening as exactly the same issue re-appeared after upgrading go-app from v9 to v10 -> https://github.com/rtrzebinski/simple-memorizer-4/pull/6/files

As a part of the v10 upgrade I made this change in 2 files:

```
internal/components/lesson_row.go
internal/components/exercise_row.go
```

old -> `app.Button().Text("Delete").OnClick(c.onDelete(c.lesson.Id), fmt.Sprintf("%p", c))`
new -> `app.Button().Text("Delete").OnClick(c.onDelete(c.lesson.Id))`

(example from `internal/components/lesson_row.go`)

Iirc I added this before exactly to fix this issue, and it worker fine, but now `OnClick()` does not accept string as a 2nd parameter.

Instead it expects -> `OnClick(h EventHandler, options ...EventOption) HTMLButton`

I see:

```go
type EventOption struct {
	name  string
	value string
}

func makeEventHandler(event string, h EventHandler, options ...EventOption) eventHandler {
	handler := eventHandler{
		event:     event,
		goHandler: h,
	}

	for _, option := range options {
		switch option.name {
		case "scope":
			handler.scope = option.value

		case "passive":
			handler.passive = true
		}
	}
	return handler
}
```

But I need help to understand how to convert previously used string to this new structure.

Thank you for any hints.
